File: app/desktop/src/main/java/dev/schlaubi/tonbrett/app/desktop/NativeUtil.java
package dev.schlaubi.tonbrett.app.desktop;

import org.jetbrains.annotations.NotNull;
import org.jetbrains.annotations.Nullable;

import java.lang.foreign.*;
import java.lang.invoke.MethodHandle;
import java.net.URI;
import java.util.Objects;

/**
 * Utility class for calling {@code uwp_helper} functions.
 * <p>
 * Implemented in Java because {@code @PolymorphicSignature} doesn't seem to work in Kotlin
 */
public class NativeUtil {

    private static final NativeUtil instance = new NativeUtil();
    private final Linker linker = Linker.nativeLinker();
    private final SymbolLookup uwpHelper = SymbolLookup.libraryLookup("uwp_helper", SegmentScope.auto());

    private final MethodHandle launchUriMethod = linker.downcallHandle(
            uwpHelper.find("launch_uri").orElseThrow(),
            FunctionDescriptor.ofVoid(ValueLayout.ADDRESS)
    );

    private final MethodHandle getAppdataFolder = linker.downcallHandle(
            uwpHelper.find("get_appdata_folder").orElseThrow(),
            FunctionDescriptor.ofVoid(ValueLayout.ADDRESS)
    );

    private final MethodHandle getAppdataFolderPathLength = linker.downcallHandle(
            uwpHelper.find("get_appdata_folder_path_length").orElseThrow(),
            FunctionDescriptor.of(ValueLayout.JAVA_LONG)
    );

    /**
     * Tries to launch the URI using the UWP {@code Launcher}.
     *
     * @param uri the {@link URI} to launch
     * @throws Throwable if an error occurs
     */
    public static void launchUri(@NotNull URI uri) throws Throwable {
        Objects.requireNonNull(uri, "uri must not be null");
        try (var arena = Arena.openConfined()) {
            var url = arena.allocateUtf8String(uri.toString());

            instance.launchUriMethod.invoke(url);
        }
    }

    private static long getAppdataFolderLength() throws Throwable {
        return (long) instance.getAppdataFolderPathLength.invoke();
    }

    /**
     * Tries to retrieve the current UWP app data folder.
     *
     * @return the absolute path to the folder
     * @throws Throwable if an error occurrs
     */
    @Nullable
    public static String getAppdataFolder() throws Throwable {
        var length = getAppdataFolderLength();
        if (length == 0) {
            return null;
        }
        try(var arena = Arena.openConfined()) {
            var buffer = arena.allocateArray(ValueLayout.JAVA_CHAR, length);
            instance.getAppdataFolder.invoke(buffer);
            return new String(buffer.toArray(ValueLayout.JAVA_CHAR));
        }
    }
}

File: app/desktop/src/main/kotlin/NativeUtil.kt
package dev.schlaubi.tonbrett.app.desktop

import dev.schlaubi.tonbrett.app.desktop.uwp_helper.AppDataRoamingResult.`is_error$get`
import dev.schlaubi.tonbrett.app.desktop.uwp_helper.AppDataRoamingResult.`length$get`
import dev.schlaubi.tonbrett.app.desktop.uwp_helper.UwpHelper.*
import java.lang.foreign.Arena
import java.net.URI

/**
 * Tries to launch the URI using the UWP `Launcher`.
 *
 * @param uri the [URI] to launch
 */
fun launchUri(uri: URI) = Arena.openConfined().use { arena ->
    val url = arena.allocateUtf8String(uri.toString())
    launch_uri(url)
}

/**
 * Tries to retrieve the current UWP app data folder.
 *
 * @return the absolute path to the folder
 */
fun getAppDataRoaming() = Arena.openConfined().use { arena ->
    val result = get_app_data_roaming(arena)
    val isError = `is_error$get`(result)
    val length = `length$get`(result).coerceAtLeast(0)
    val buffer = arena.allocateArray(uint16_t, length)
    copy_string_from_get_app_data_roaming_result_into_buffer(result, buffer)
    val shortArray = buffer.toArray(uint16_t)
    val charArray = CharArray(shortArray.size)
    for ((index, short) in shortArray.withIndex()) {
        charArray[index] = short.toInt().toChar()
    }
    val string = String(charArray)
    if (isError) throw Exception(string) else string
}
